### ISSUE TITLE
False-positive: Unnecessary type casting inspection for multi-dimensional array

### ISSUE THREAD

**Issue by tfettig01** [kalessil/phpinspectionsea#630]
The first has the false positive. The rest of the three are good.
```php
function bar1(array $response): int
{
    // False-positive: Unnecessary type casting inspection
    return (int)($response['bar']['a'] ?? $response['bar']['b'] ?? 1);
}
function baz1(array $response): int
{
    // Good
    return (int)($response['bar']['a'] ?? $response['bar']['b'] ?? '1');
}
function bar2(array $response): int
{
    // Good
    return (int)($response['a'] ?? $response['b'] ?? 1);
}
function baz2(array $response): int
{
    // Good
    return (int)($response['a'] ?? $response['b'] ?? '1');
}
```

Originally reported at https://bitbucket.org/kalessil/phpinspectionsea/issues/310/false-positive-unnecessary-type-casting

**Comment by kalessil** [comment]
Ok, this is something I need to report to JB:
- `$x = $response['a'];`: type resolved to mixed (`array $response`)
- `$x = $response['a'][b];`: type resolved is empty (`array $response`)

Meanwhile PhpDoc, eg. `@param string[][] $response` would be a feasible workaround to feed additional types info into IDE

**Comment by tfettig01** [comment]
Awesome, Now I just need to remember where in my codebase I found this.

**Comment by kalessil** [comment]
Done, reported here: https://youtrack.jetbrains.com/issue/WI-38711
